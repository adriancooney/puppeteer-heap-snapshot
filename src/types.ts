export type HeapSnapshotNode = Uint32Array;
export type HeapSnapshotEdge = Uint32Array;

export type HeapSnapshotStructuredEdge = {
  id: number;
  type: string;
  name?: string;
  index?: number;
  nodeId: number;
};

export type HeapSnapshotStructuredNode = {
  id: number;
  nodeIndex: number;
  type: string;
  name: string;
  size: number;
  edgeCount: number;
  traceNodeId: number;
  detachness: number;
  edgeIds: number[];
};

export type HeapSnapshotStructuredGraph = {
  node: HeapSnapshotStructuredNode;
  edges: {
    isCircular: boolean;
    edge: HeapSnapshotStructuredEdge;
    graph: HeapSnapshotStructuredGraph | null;
  }[];
};

export type MetaValue = string | string[];
export type SerializedHeapSnapshot = {
  snapshot: {
    node_count: number;
    edge_count: number;
    trace_function_count: number;

    meta: {
      node_fields: MetaValue[];
      node_types: MetaValue[];
      edge_fields: MetaValue[];
      edge_types: MetaValue[];
      trace_function_info_fields: MetaValue[];
      trace_node_fields: MetaValue[];
      sample_fields: MetaValue[];
      location_fields: MetaValue[];
    };
  };

  nodes: number[];
  edges: number[];
  strings: string[];

  trace_function_infos: number[];
  samples: number[];
  trace_tree: number[];
  locations: number[];
};

export type HeapSnapshot = Omit<SerializedHeapSnapshot, "nodes" | "edges"> & {
  edges: Uint32Array;
  nodes: Uint32Array;
};

export type BuiltHeapValue =
  | undefined
  | null
  | string
  | number
  | RegExp
  | boolean
  | BuiltHeapValue[]
  | {
      [key: string]: BuiltHeapValue;
    };
