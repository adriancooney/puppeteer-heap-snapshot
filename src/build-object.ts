import { debug } from "./util";
import {
  BuiltHeapValue,
  HeapSnapshot,
  HeapSnapshotStructuredGraph,
  HeapSnapshotStructuredEdge,
} from "./types";
import {
  createStructuredGraph,
  createStructuredNode,
} from "./structured-graph";

type BuildObjectFromNodeIdOption = {
  propertyFilter?: (propertyName: string) => boolean;
  maxDepth?: number;
  unwantedNodeNames?: readonly string[];
};

export function buildObjectFromNodeId(
  heapSnapshot: HeapSnapshot,
  nodeId: number,
  options: BuildObjectFromNodeIdOption
): BuiltHeapValue {
  debug(`building node object for node ${nodeId}`);

  const { propertyFilter = () => true, maxDepth, unwantedNodeNames } = options;

  // Get structuredNode early to figure out if we should return quickly in case of unwanted node name
  const structuredNode = createStructuredNode(heapSnapshot, nodeId);

  if (unwantedNodeNames?.includes(structuredNode.name)) {
    debug(`found node with unwanted name ${structuredNode.name}, skipping...`);
    return;
  }

  const graph = createStructuredGraph(heapSnapshot, nodeId, structuredNode, {
    maxDepth,
    edgeFilter: (edge: HeapSnapshotStructuredEdge) => {
      if (edge.type === "property") {
        return edge.name! !== "__proto__" && propertyFilter(edge.name!);
      }

      return (
        edge.type === "element" ||
        edge.name === "value" ||
        edge.type === "hidden"
      );
    },
  });

  if (graph.node.type !== "object") {
    throw new Error(`Node '${nodeId}' is not object, cannot build object`);
  }

  debug(`compiling graph node object ${nodeId}`);
  return compileGraphNodeObject(graph);
}

function compileGraphNodeObject(
  graph: HeapSnapshotStructuredGraph
): BuiltHeapValue {
  const node = graph.node;
  const nodeType = node.type;
  const nodeName = node.name;

  const edges = graph.edges.filter(({ graph }) => {
    const edgeNodeType = graph?.node?.type;

    if (!edgeNodeType) {
      return false;
    }

    if (edgeNodeType === "object") {
      return ["Array", "Object"].includes(graph.node.name);
    } else {
      return (
        ["array", "string", "number", "regexp"].includes(edgeNodeType) ||
        isBoolean(graph) ||
        isNull(graph)
      );
    }
  });

  if (nodeType === "array") {
    return edges.map(({ graph: arrayValueGraph }) =>
      compileGraphNodeObject(arrayValueGraph!)
    );
  } else if (nodeType === "object") {
    if (nodeName === "Object") {
      return edges.reduce(
        (obj, { edge, graph }) => ({
          ...obj,
          [edge.name!]: compileGraphNodeObject(graph!),
        }),
        {}
      );
    } else if (nodeName === "Array") {
      return compileGraphNodeObject({
        ...graph,
        node: {
          ...graph.node,
          type: "array",
        },
      });
    } else {
      throw new Error(`Unknown or unsupported object with type '${nodeName}'`);
    }
  } else if (nodeType === "string") {
    return node.name;
  } else if (nodeType === "regexp") {
    return new RegExp(node.name as string);
  } else if (nodeType === "number") {
    return parseFloat(
      edges.find(({ edge }) => edge.name === "value")!.graph!.node!.name
    );
  } else if (isBoolean(graph)) {
    return (
      graph.edges.find(({ graph }) => graph?.node.type === "string")?.graph
        ?.node.name === "true"
    );
  } else if (isNull(graph)) {
    return null;
  } else {
    throw new Error(
      `Unknown graph node type '${nodeType}', unable to compile graph object`
    );
  }
}

function isBoolean(graph: HeapSnapshotStructuredGraph): boolean {
  return (
    graph.node.type === "hidden" &&
    !!graph.edges.find(({ graph }) => graph?.node.name === "boolean")
  );
}

function isNull(graph: HeapSnapshotStructuredGraph): boolean {
  return (
    graph.node.type === "hidden" &&
    !!graph.edges.find(({ graph }) => graph?.node.name === "object") &&
    !!graph.edges.find(({ graph }) => graph?.node.name === "null")
  );
}
