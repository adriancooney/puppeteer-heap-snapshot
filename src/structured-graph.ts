import {
  findEdgeById,
  findNodeById,
  getFieldValue,
  getNodeAtIndex,
  getNodeEdgeIds,
} from "./snapshot";
import {
  HeapSnapshot,
  HeapSnapshotStructuredEdge,
  HeapSnapshotStructuredGraph,
  HeapSnapshotStructuredNode,
} from "./types";

export function createStructuredGraph(
  heapSnapshot: HeapSnapshot,
  nodeId: number,
  structuredNode: HeapSnapshotStructuredNode,
  {
    maxDepth = Infinity,
    edgeFilter = () => true,
  }: {
    maxDepth?: number;
    edgeFilter?: (edge: HeapSnapshotStructuredEdge) => boolean;
  } = {},
  nodeIdStack: number[] = []
): HeapSnapshotStructuredGraph {
  const structuredEdges = structuredNode.edgeIds
    .map((edgeId) => createStructuredEdge(heapSnapshot, edgeId))
    .filter(edgeFilter);

  return {
    node: structuredNode,
    edges: structuredEdges.map((structuredEdge) => {
      const isCircular = nodeIdStack.includes(structuredEdge.nodeId);

      return {
        isCircular,
        edge: structuredEdge,
        graph:
          !isCircular && nodeIdStack.length < maxDepth
            ? createStructuredGraph(
                heapSnapshot,
                structuredEdge.nodeId,
                createStructuredNode(heapSnapshot, structuredEdge.nodeId),
                { maxDepth, edgeFilter },
                [...nodeIdStack, nodeId]
              )
            : null,
      };
    }),
  };
}

export function createStructuredNode(
  heapSnapshot: HeapSnapshot,
  nodeId: number
): HeapSnapshotStructuredNode {
  const { node, nodeIndex } = findNodeById(heapSnapshot, nodeId);
  const edgeIds = getNodeEdgeIds(heapSnapshot, nodeId);

  return {
    nodeIndex,
    edgeIds,
    edgeCount: getFieldValue(
      heapSnapshot,
      "node",
      "edge_count",
      node
    ) as number,
    type: getFieldValue(heapSnapshot, "node", "type", node) as string,
    name: getFieldValue(heapSnapshot, "node", "name", node) as string,
    id: getFieldValue(heapSnapshot, "node", "id", node) as number,
    size: getFieldValue(heapSnapshot, "node", "self_size", node) as number,
    traceNodeId: getFieldValue(
      heapSnapshot,
      "node",
      "trace_node_id",
      node
    ) as number,
    detachness: getFieldValue(
      heapSnapshot,
      "node",
      "detachedness",
      node
    ) as number,
  };
}

function createStructuredEdge(
  heapSnapshot: HeapSnapshot,
  edgeId: number
): HeapSnapshotStructuredEdge {
  const { edge } = findEdgeById(heapSnapshot, edgeId);
  const type = getFieldValue(heapSnapshot, "edge", "type", edge) as string;
  // to_node is the index of the first value of the node in the nodes array
  const nodeIndex =
    (getFieldValue(heapSnapshot, "edge", "to_node", edge) as number) /
    heapSnapshot.snapshot.meta.node_fields.length;
  const nodeId = getFieldValue(
    heapSnapshot,
    "node",
    "id",
    getNodeAtIndex(heapSnapshot, nodeIndex)
  ) as number;
  const isIndex = type === "element" || type === "hidden";
  const nameOrIndex = getFieldValue(
    heapSnapshot,
    "edge",
    "name_or_index",
    edge,
    !isIndex
  );

  return {
    id: edgeId,
    type,
    nodeId,
    name: !isIndex ? (nameOrIndex as string) : undefined,
    index: isIndex ? (nameOrIndex as number) : undefined,
  };
}

export function formatStructuredGraph(
  structuredGraph: HeapSnapshotStructuredGraph,
  indentSize = 0
): string {
  const lines = [];
  const indent = "   ".repeat(indentSize);
  const { node, edges } = structuredGraph;

  lines.push(
    indent +
      `Node{${node.id}} ${node.name} (type: ${node.type}, size: ${node.size})`
  );

  edges.forEach(({ edge, graph, isCircular }) => {
    lines.push(
      indent +
        `-> Edge{${edge.id}} ${edge.name || edge.index} (${
          edge.type
        }) -> Node{${edge.nodeId}}` +
        (isCircular ? " (circular)" : "")
    );

    if (graph) {
      lines.push(formatStructuredGraph(graph, indentSize + 1));
    }
  });

  return lines.join("\n");
}
