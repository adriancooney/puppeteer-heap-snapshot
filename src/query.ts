import { debug } from "./util";
import { BuiltHeapValue, HeapSnapshot } from "./types";
import { findNodeIdsWithProperties } from "./snapshot";
import { buildObjectFromNodeId } from "./build-object";

type FindObjectOptions = {
  ignoreProperties?: readonly string[];
  maxDepth?: number;
  unwantedNodeNames?: readonly string[];
};

type FindObjectReturnValue<
  Props extends readonly string[],
  Options extends FindObjectOptions
> = {
  [key in Exclude<Props[number], Options["ignoreProperties"]>]: BuiltHeapValue;
};

export function findObjectsWithProperties<
  Props extends readonly string[],
  Options extends FindObjectOptions
>(
  heapSnapshot: HeapSnapshot,
  properties: Props,
  options?: Options
): FindObjectReturnValue<Props, Options>[] {
  debug("finding objects with properties", properties, options);

  const nodeIds = findNodeIdsWithProperties(heapSnapshot, properties);

  debug(`${nodeIds.length} nodes found, compiling objects`);

  if (nodeIds.length > 5) {
    debug(
      `more than 5 nodes found, this may be slow - to improve performance, increase the specifity of your query or ignore unwanted properties on the target object`
    );
  }

  return nodeIds
    .map((nodeId) => {
      debug(`node ${nodeId} found`);
      return buildObjectFromNodeId(heapSnapshot, nodeId, {
        ...options,
        propertyFilter: options?.ignoreProperties
          ? (prop) => !options.ignoreProperties!.includes(prop)
          : undefined,
      });
    })
    .filter((v) => v !== undefined) as FindObjectReturnValue<Props, Options>[];
}

export function findObjectWithProperties<
  Props extends readonly string[],
  Options extends FindObjectOptions
>(
  heapSnapshot: HeapSnapshot,
  properties: Props,
  options?: Options
): FindObjectReturnValue<Props, Options> {
  const objects = findObjectsWithProperties<Props, Options>(
    heapSnapshot,
    properties,
    options
  );

  if (!objects.length) {
    throw new Error(
      `No object found with properties: ${properties.join(", ")}`
    );
  }

  if (objects.length > 1) {
    throw new Error(
      `More than one object found with properties: ${properties.join(", ")}`
    );
  }

  return objects[0]!;
}
