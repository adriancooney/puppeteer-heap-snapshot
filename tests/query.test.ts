import { BuiltHeapValue, HeapSnapshot } from "../src";
import { readHeapSnapshot } from "../src/fs";
import {
  findObjectsWithProperties,
  findObjectWithProperties,
} from "../src/query";

describe("query", () => {
  let heapSnapshot: HeapSnapshot;

  beforeAll(async () => {
    heapSnapshot = await readHeapSnapshot(
      __dirname + "/fixtures/data/example.heapsnapshot"
    );
  });

  describe("findObjectsWithProperties", () => {
    it("finds all objects", async () => {
      expect(
        sortObjects(
          findObjectsWithProperties(heapSnapshot, ["bar", "id"] as const)
        )
      ).toMatchInlineSnapshot(`
        Array [
          Object {
            "bar": 2,
            "foo": 1,
            "id": "my-object-1",
          },
          Object {
            "bar": 1,
            "boot": 2,
            "id": "my-object-2",
          },
        ]
      `);
    });

    it("finds some objects", async () => {
      expect(findObjectsWithProperties(heapSnapshot, ["foo"]))
        .toMatchInlineSnapshot(`
        Array [
          Object {
            "bar": 2,
            "foo": 1,
            "id": "my-object-1",
          },
          Object {
            "foo": Object {},
          },
        ]
      `);
    });

    it("finds some objects respecting unwantedNodeNames and maxDepth options", async () => {
      const heapSnapshot = await readHeapSnapshot(
        __dirname + "/fixtures/data/example-with-builtins.heapsnapshot"
      );

      expect(
        findObjectsWithProperties(heapSnapshot, ["href"], {
          unwantedNodeNames: [
            "Location",
            "HTMLAnchorElement",
            "URL",
            "HTMLLinkElement",
            "SVGGradientElement",
            "SVGFilterElement",
          ],

          maxDepth: 1,
        })
      ).toMatchSnapshot();
    });

    it("finds some objects with more specific properties", async () => {
      expect(findObjectsWithProperties(heapSnapshot, ["bar", "boot"]))
        .toMatchInlineSnapshot(`
        Array [
          Object {
            "bar": 1,
            "boot": 2,
            "id": "my-object-2",
          },
        ]
      `);
    });

    it("finds no objects", async () => {
      expect(
        findObjectsWithProperties(heapSnapshot, ["foo", "boot"])
      ).toMatchInlineSnapshot(`Array []`);
    });

    it("ignores properties", async () => {
      expect(
        findObjectsWithProperties(heapSnapshot, ["bar"], {
          ignoreProperties: ["foo"],
        })
      ).toMatchInlineSnapshot(`
        Array [
          Object {
            "bar": 1,
            "boot": 2,
            "id": "my-object-2",
          },
          Object {
            "bar": 2,
            "id": "my-object-1",
          },
        ]
      `);
    });

    it("parses complex objects", async () => {
      expect(findObjectsWithProperties(heapSnapshot, ["myComplexObject"]))
        .toMatchInlineSnapshot(`
        Array [
          Object {
            "b": null,
            "c": true,
            "circularRef": Object {
              "foo": Object {},
            },
            "d": Infinity,
            "g": Object {
              "a": 1,
              "b": 2,
              "c": Object {
                "a": 1,
                "b": 2,
              },
            },
            "h": 0.1,
            "i": 1e-12,
            "j": "foobar",
            "k": /foo\\.\\*b\\[\\^"\\]\\+a\\(r\\)\\?/,
            "myComplexObject": 1,
          },
        ]
      `);
    });
  });

  describe("findObjectWithProperties", () => {
    it("finds a single object", () => {
      expect(findObjectWithProperties(heapSnapshot, ["bar", "boot"]))
        .toMatchInlineSnapshot(`
        Object {
          "bar": 1,
          "boot": 2,
          "id": "my-object-2",
        }
      `);
    });

    it("throws if no object is found", () => {
      expect(() =>
        findObjectWithProperties(heapSnapshot, ["foo", "boot"])
      ).toThrowErrorMatchingInlineSnapshot(
        `"No object found with properties: foo, boot"`
      );
    });

    it("throws if more than one object found", () => {
      expect(() =>
        findObjectWithProperties(heapSnapshot, ["bar"])
      ).toThrowErrorMatchingInlineSnapshot(
        `"More than one object found with properties: bar"`
      );
    });
  });
});

function sortObjects<T>(objects: (T & { id: BuiltHeapValue })[]): T[] {
  return objects.sort((a, b) => (a.id as string).localeCompare(b.id as string));
}
