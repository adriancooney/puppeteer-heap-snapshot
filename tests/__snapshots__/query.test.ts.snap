// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query findObjectsWithProperties finds some objects respecting unwantedNodeNames and maxDepth options 1`] = `
Array [
  Object {
    "children": "disclaimer",
    "href": "/disclaimer/",
  },
  Object {
    "children": "privacy policy",
    "href": "/privacy/",
  },
  Object {
    "hash": "",
    "host": "polypane.app",
    "hostname": "polypane.app",
    "href": "https://polypane.app/css-specificity-calculator/",
    "key": "initial",
    "origin": "https://polypane.app",
    "pathname": "/css-specificity-calculator/",
    "port": "",
    "protocol": "https:",
    "search": "",
  },
  Object {
    "children": "disclaimer",
    "href": "/disclaimer/",
  },
  Object {
    "children": Array [],
    "className": "Header-styles-module--logo--1JkU1",
    "href": "/",
  },
  Object {
    "children": "Skip to footer",
    "className": "SkipLink-styles-module--skiplink--2HGNv",
    "href": "#footer",
  },
  Object {
    "children": "Skip to content",
    "className": "SkipLink-styles-module--skiplink--2HGNv",
    "href": "#main",
  },
  Object {},
  Object {},
  Object {},
  Object {},
  Object {
    "children": "Responsive design glossary",
    "href": "/responsive-design-glossary/",
  },
  Object {
    "children": "Create Polypane workspace",
    "href": "/create-workspace/",
  },
  Object {
    "children": "Your screen size wall",
    "href": "/responsive-website-test/",
  },
  Object {
    "aria-current": "page",
    "children": "CSS specificity calculator",
    "className": "",
    "href": "/css-specificity-calculator/",
    "style": Object {},
  },
  Object {
    "children": "Color contrast checker",
    "href": "/color-contrast/",
  },
  Object {
    "children": "Pricing",
    "href": "/pricing/",
  },
  Object {
    "children": "For QA",
    "href": "/quality-assurance/",
  },
  Object {
    "children": "For Agencies",
    "href": "/agencies/",
  },
  Object {
    "children": "For Marketers",
    "href": "/marketers/",
  },
  Object {
    "children": "All free tools",
    "href": "/resources/",
  },
  Object {
    "children": "Home",
    "href": "/",
  },
  Object {
    "children": "About",
    "href": "/about/",
  },
  Object {
    "children": "Blog",
    "href": "/blog/",
  },
  Object {
    "children": "Newsletter",
    "href": "/newsletter/",
  },
  Object {
    "children": "Accolades",
    "href": "/accolades/",
  },
  Object {
    "children": "Changelog",
    "href": "/docs/changelog/",
  },
  Object {
    "children": "Testimonials",
    "href": "/testimonials/",
  },
  Object {
    "children": "Download",
    "href": "/download/",
  },
  Object {
    "children": "Integrations",
    "href": "/integrations/",
  },
  Object {
    "children": "Docs",
    "href": "/docs/",
  },
  Object {
    "children": "Site Quality",
    "href": "/site-quality/",
  },
  Object {
    "children": "Accessibility",
    "href": "/accessibility/",
  },
  Object {
    "children": "Accessibility Statement",
    "href": "/accessibility-statement/",
  },
  Object {
    "children": "Disclaimer",
    "href": "/disclaimer/",
  },
  Object {
    "children": "Legal",
    "href": "/legal/",
  },
  Object {
    "children": Array [],
    "className": "Releases-styles-module--fulllink--3XjeX",
    "href": "/docs/changelog/",
  },
  Object {
    "children": Array [],
    "href": "/blog/polypane-8-1-resizable-element-tree-disable-js-feature-new-debug-tools-and-more/",
  },
  Object {
    "children": Array [],
    "href": "/blog/polypane-8-1-resizable-element-tree-disable-js-feature-new-debug-tools-and-more/",
  },
  Object {
    "children": "Blog",
    "href": "/blog/",
  },
  Object {
    "children": "Testimonials",
    "href": "/testimonials/",
  },
  Object {
    "children": "Docs",
    "href": "/docs/",
  },
  Object {
    "children": Array [],
    "href": "/create-workspace/",
  },
  Object {
    "children": Array [],
    "href": "/visualize-browser-sizes/",
  },
  Object {
    "aria-current": "page",
    "children": Array [],
    "className": "",
    "href": "/css-specificity-calculator/",
    "style": Object {},
  },
  Object {
    "children": Array [],
    "href": "/color-contrast/",
  },
  Object {
    "children": "Integrations",
    "href": "/integrations/",
  },
  Object {
    "children": Array [],
    "href": "/agencies/",
  },
  Object {
    "children": Array [],
    "href": "/marketers/",
  },
  Object {
    "children": Array [],
    "href": "/quality-assurance/",
  },
  Object {
    "children": Array [],
    "href": "/site-quality/",
  },
  Object {
    "children": Array [],
    "href": "/accessibility/",
  },
  Object {
    "children": Array [],
    "href": "/blog/polypane-9-screenshot-editor-structured-data-support-and-new-debug-tools/",
  },
  Object {
    "children": Array [],
    "className": "CTA-styles-module--button--2bMFr",
    "href": "https://dashboard.polypane.app/register/",
  },
  Object {
    "children": Object {},
    "className": "Header-styles-module--button--3nyub",
    "href": "https://dashboard.polypane.app/register/",
  },
  Object {
    "children": "Pricing",
    "href": "/pricing/",
  },
  Object {
    "children": Array [],
    "href": "/product-tour/?ref=designers",
  },
  Object {
    "children": Array [],
    "href": "/product-tour/?ref=developers",
  },
  Object {
    "children": Array [],
    "href": "/product-tour/?ref=responsive-design",
  },
  Object {
    "children": Array [],
    "href": "/product-tour/",
  },
  Object {
    "activeStyle": Object {},
    "children": "For Designers",
    "href": "/product-tour/?ref=designers",
  },
  Object {
    "activeStyle": Object {},
    "children": "For Developers",
    "href": "/product-tour/?ref=developers",
  },
  Object {
    "children": "Manage Account",
    "href": "https://dashboard.polypane.app/",
  },
  Object {
    "children": "Support",
    "href": "/support/",
  },
  Object {
    "children": "privacy policy",
    "href": "/privacy/",
  },
  Object {
    "children": Array [],
    "href": "/product-tour/?ref=developers",
  },
  Object {
    "children": Array [],
    "href": "/product-tour/?ref=responsive-design",
  },
  Object {
    "children": Array [],
    "href": "/product-tour/?ref=designers",
  },
  Object {
    "children": "Privacy",
    "href": "/privacy/",
  },
  Object {
    "children": Object {},
    "className": "Header-styles-module--button--3nyub",
    "href": "https://dashboard.polypane.app/register/",
  },
  Object {
    "href": "https://polypane.app/rss.xml",
    "rel": "alternate",
    "title": "Polypane Blog",
    "type": "application/rss+xml",
  },
  Object {
    "href": "https://chrome.google.com/webstore/detail/eofbapfmbfmpeplodnehlkkgpkklmapp",
    "rel": "chrome-webstore-item",
  },
  Object {
    "as": "font",
    "crossOrigin": "anonymous",
    "href": "/static/Inter-roman.var-subset-4205477ad91e63d6d092b4e155e529a2.woff2",
    "rel": "preload",
    "type": "font/woff2",
  },
  Object {
    "as": "font",
    "crossOrigin": "anonymous",
    "href": "/static/Inter-italic.var-subset-b850defff6a5dd08ab5b60d16fd28a87.woff2",
    "rel": "preload",
    "type": "font/woff2",
  },
  Object {
    "children": Array [],
    "className": "CTA-styles-module--button--2bMFr",
    "href": "https://dashboard.polypane.app/register/",
  },
  Object {
    "href": "https://polypane.app/rss.xml",
    "rel": "alternate",
    "title": "Polypane Blog",
    "type": "application/rss+xml",
  },
  Object {
    "href": "https://chrome.google.com/webstore/detail/eofbapfmbfmpeplodnehlkkgpkklmapp",
    "rel": "chrome-webstore-item",
  },
  Object {
    "as": "font",
    "crossOrigin": "anonymous",
    "href": "/static/Inter-roman.var-subset-4205477ad91e63d6d092b4e155e529a2.woff2",
    "rel": "preload",
    "type": "font/woff2",
  },
  Object {
    "as": "font",
    "crossOrigin": "anonymous",
    "href": "/static/Inter-italic.var-subset-b850defff6a5dd08ab5b60d16fd28a87.woff2",
    "rel": "preload",
    "type": "font/woff2",
  },
  Object {
    "children": Array [],
    "href": "https://twitter.com/intent/user?screen_name=polypane",
  },
  Object {
    "children": Array [],
    "href": "https://polypane-slack-invite.herokuapp.com/",
  },
  Object {
    "activeStyle": Object {},
    "children": "Responsive design",
    "href": "/product-tour/?ref=responsive-design",
  },
  Object {
    "activeStyle": Object {},
    "children": "For Developers",
    "href": "/product-tour/?ref=developers",
  },
  Object {
    "children": "Manage Account",
    "href": "https://dashboard.polypane.app/",
  },
  Object {
    "activeStyle": Object {},
    "children": "For Designers",
    "href": "/product-tour/?ref=designers",
  },
  Object {
    "activeStyle": Object {},
    "children": "Responsive design",
    "href": "/product-tour/?ref=responsive-design",
  },
  Object {
    "children": Array [],
    "href": "https://twitter.com/intent/user?screen_name=polypane",
  },
  Object {
    "children": Array [],
    "href": "https://polypane-slack-invite.herokuapp.com/",
  },
  Object {
    "hash": "",
    "host": "polypane.app",
    "hostname": "polypane.app",
    "href": "https://polypane.app/css-specificity-calculator/",
    "key": "initial",
    "origin": "https://polypane.app",
    "pathname": "/css-specificity-calculator/",
    "port": "",
    "protocol": "https:",
    "search": "",
  },
  Object {},
]
`;
